blog/posts/2019-10-27-unit-testing-private-non-exported-functions-with-rewire.md

```diff
@@ -0,0 +1,89 @@
+---
+authorId: rwwagner90
+categories: 
+  - testing
+date: '2019-10-27'
+nextSlug: shepherd-popper-to-tether
+nextTitle: 'There and Back Again - Popper to Tether'
+previousSlug: ember-data-passing-query-params-to-save
+previousTitle: Ember Data | Passing query params to .save()
+slug: unit-testing-private-non-exported-functions-with-rewire
+title: Unit Testing Private, Non-exported Functions with Rewire
+---
+
+When a JavaScript function is exported, it is straightforward to unit test. 
+We can import it directly into our test and test the functionality. For example,
+we could use something like Jest to setup a simple import and unit test.
+
+## Exported
+
+```js
+// exported.js
+
+export function foo() {
+  return 'bar';
+}
+```
+
+```js
+// exported.spec.js
+
+import { foo } from './exported.js'
+
+describe('Exported', () => {
+  describe('foo', () => {
+    expect(foo()).toBe('bar');
+  });
+});
+```
+
+## Private / Non-exported
+
+How would we test functions that we do not expose via `export` though?
+
+There is a great tool, [rewire](https://www.npmjs.com/package/rewire), that allows getting references 
+to functions that are not explicitly exported. Assuming you are using Babel, we can use 
+[babel-plugin-rewire](https://github.com/speedskater/babel-plugin-rewire) to make the setup more simple.
+
+```bash
+  npm install babel-plugin-rewire --save-dev
+```
+
+Then we need to add it to our plugins for testing, in our `babel.config.js`.
+
+```js
+...
+plugins: [
+  'babel-plugin-rewire',
+  ...
+]
+...
+```
+
+We can then import the whole file, and get references to the non-exported functions, using
+`__get__`.
+
+```js
+// private.js
+
+function _foo() {
+  return 'bar';
+}
+```
+
+```js
+// private.spec.js
+
+import Private from './private.js'
```

Comment:
could you add a short reference to what this `Private` default is and anything else rewire does?

Reply:
This is just a random name. Perhaps that is not clear, but we could call it `Foo`, `Bar`, etc